### ISSUE TITLE
Feature improvement request: include routes into component hierarchy graph

### ISSUE THREAD

**Issue by MShekow** [CoderAllan/vscode-angulartools#280]
Hi Allan, your plug-in really helps to visualize dependencies in Angular projects. I particularly like the *component hierarchy graph*. I noticed, however, that if a component internally uses other components *indirectly*, via having a `<router-outlet></router-outlet>` in its HTML template (and routes defined in, say, the `AppRoutingModule`), then this dependency is missing from the component hierarchy graph.

Do you see a way to also include these indirect dependencies?

**Comment by CoderAllan** [comment]
To be honest I'm not a hardcore Angular developer primarily do backend stuff in C# and I'm not that familiar with the routing in Angular so I need some more detail of how these relationships should be detected. 
This is one of the example apps I have used to test the extension with: https://github.com/DanWahlin/Angular-JumpStart
For example in that app there is the app-routing-module.ts that define the Route for customers that uses the CustomersModule that uses the CustomerComponent as default and the app.component.html has a router-outlet defined. So is the relationship you would like to have visualized is from the app.component to the customer.component in the component hierarchy graph? 
Something like this:
<img width="489" alt="image" src="https://user-images.githubusercontent.com/11722431/142570283-6220b015-aafc-45c5-bce9-c51bec75e884.png">

**Comment by MShekow** [comment]
Yes, correct. For the `Angular JumpStart` app, I would like to see relations such as 

`AppComponent => CustomersComponent`
`AppComponent => OrdersComponent`
`AppComponent => AboutComponent`
`CustomersComponent => CustomerOrdersComponent`
`CustomersComponent => CustomerDetailsComponent`
`CustomersComponent => CustomerEditComponent`

The general parsing logic would be like this:
- For every component `c` whose HTML template contains a `<router-outlet></router-outlet>`:
  - Get the corresponding _module_ `m` that references `c`
  - Check the _imports_ or _declarations_ of `m`'s `@NgModule` for modules which have `RouterModule.forRoot(someRoutes)` defined
  - Parse `someRoutes`, and find other components `c'` that are referenced in `someRoutes` (see the `component` key, or `loadChildren()` key of each route)
  - For every `c'` you find, there is a relation `c => c'` in the component hierarchy graph

**Comment by CoderAllan** [comment]
@MShekow I have now published version 1.9.0 of the extension that is rendering the routing relations to components referenced directly in the routing modules, like shown below:
<img width="355" alt="image" src="https://user-images.githubusercontent.com/11722431/145517152-a7059c53-240e-44ae-9019-b71b0a90bfb4.png">
When routes are defined as references to module(like in the screenshot below) I need to implement some kind of parsing in order to discover the components that are referenced. I will be working on to solve that issue so all routing relation can be rendered.
<img width="729" alt="image" src="https://user-images.githubusercontent.com/11722431/145517285-c3c9c0c5-7328-4497-ac59-ba859adc13bf.png">
I will leave this issue open until I get the final part of the rendering implemented.

**Comment by MShekow** [comment]
Dear Allan, thanks for implementing some of the functionality :). With v1.9.0, I do see new orange-colored arrows in the component hierarchy graph.

One minor problem is that the origin of the arrow is not correct, at least for my own project. In my project, the component hierarchy graph claims that my `AppComponent` (the one referenced in `index.html`) routes to the other components that were discovered in the `AppRoutingModule`. In reality, however, `AppComponent` includes another component `X`, and the `<router-outlet></router-outlet>` is present in `X`'s HTML template. Therefore, the origin of the orange arrows should be `X`, not `AppComponent`.

**Comment by CoderAllan** [comment]
Thanks for the feedback. Sounds like i have made an assumption that didn't hold...

I have created a fork of Angular-Jumpstart: https://github.com/CoderAllan/Angular-JumpStart
Can you pull that repository, create/change some of components to illustrate the issue, then create a pull request so I can see exactly how your component hierarchy looks like in this situation?
